aptos-move/framework/aptos-framework/sources/voting.spec.move

```diff
@@ -0,0 +1,223 @@
+spec aptos_framework::voting {
+    spec module {
+        pragma verify = true;
+        pragma aborts_if_is_strict;
+    }
+
+    spec register<ProposalType: store>(account: &signer) {
+        let addr = signer::address_of(account);
+
+        // Will abort if there's already a `VotingForum<ProposalType>` under addr
+        aborts_if exists<VotingForum<ProposalType>>(addr);
+        // Creation of 4 new event handles changes the account's `guid_creation_num`
+        aborts_if !exists<account::Account>(addr);
+        let register_account = global<account::Account>(addr);
+        aborts_if register_account.guid_creation_num + 4 > MAX_U64;
+        // `type_info::type_of()` may abort if the type parameter is not a struct
+        aborts_if !type_info::spec_is_struct<ProposalType>();
+
+        ensures exists<VotingForum<ProposalType>>(addr);
+    }
+
+    spec create_proposal<ProposalType: store>(
+        proposer: address,
+        voting_forum_address: address,
+        execution_content: ProposalType,
+        execution_hash: vector<u8>,
+        min_vote_threshold: u128,
+        expiration_secs: u64,
+        early_resolution_vote_threshold: Option<u128>,
+        metadata: SimpleMap<String, vector<u8>>,
+    ): u64 {
+        include CreateProposal<ProposalType>{is_multi_step_proposal: false};
+    }
+
+    /// The min_vote_threshold lower thanearly_resolution_vote_threshold.
+    /// Make sure the execution script's hash is not empty.
+    /// VotingForum<ProposalType> existed under the voting_forum_address.
+    /// The next_proposal_id in VotingForum is up to MAX_U64.
+    /// CurrentTimeMicroseconds existed under the @aptos_framework.
+    spec create_proposal_v2<ProposalType: store>(
+        proposer: address,
+        voting_forum_address: address,
+        execution_content: ProposalType,
+        execution_hash: vector<u8>,
+        min_vote_threshold: u128,
+        expiration_secs: u64,
+        early_resolution_vote_threshold: Option<u128>,
+        metadata: SimpleMap<String, vector<u8>>,
+        is_multi_step_proposal: bool,
+    ): u64 {
+        include CreateProposal<ProposalType>;
+    }
+
+    spec schema CreateProposal<ProposalType> {
```

Comment:
Nice to make it more modular using the schema!

Reply:
Minor comment for a convention. Could you rename this schema to `CreateProposalAbortsIf` because it's mostly about `aborts_if` without `ensures`.